### PR TITLE
Elem::orient(), MeshTools::Modification::orient_elements()

### LINKED ISSUE #3424: Add routine to fix negative jacobian for all types of elements
Negative jacobians in elements can happen for a variety of reasons, one being careless moving around of nodes (for example during a symmetry transform of a mesh, or an extrusion in the wrong direction). It's not illegitimate to want the mesh to be able to repair itself so it can be used. 

It needs to:
- swap some nodes
- update the neighbors

I think this will need to be a routine attribute on the element class

Any idea of other things to do on this?
@dschwen fyi

### PR DESCRIPTION
**Pull request by roystgnr**
This ought to close #3424

@GiudGiud - you've got a MOOSE MeshGenerator test case you can try this out on, make sure it handles bad extrusion vectors properly, before we merge here?

### CODE REVIEW

**Comment by GiudGiud** [comment]
Sure. I have the symmetry and the extrusion to try. I ll report back

File: src/mesh/mesh_modification.C [hunk 1]
```diff
@@ -186,6 +186,23 @@ void MeshTools::Modification::permute_elements(MeshBase & mesh)
...
+void MeshTools::Modification::orient_elements(MeshBase & mesh)
+{
+  LOG_SCOPE("orient_elements()", "MeshTools::Modification");
+
+  // We don't yet support doing orient() on a parent element, which
+  // would require us to consistently orient all its children and
+  // give them different local child numbers.
+  unsigned int n_levels = MeshTools::n_levels(mesh);
+  if (n_levels > 1)
+    libmesh_error();
```
Comment (GiudGiud) [inline src/mesh/mesh_modification.C:198]: so if you leave this without a message we'll need to catch it upstream in moose

**Comment by GiudGiud** [comment]
The symmetry generator works well with the set and gives the right solution on a diffusion problem with QUADs.

The (advanced)extruder returns different sidelists with this flipping compared to the manual flipping logic we had.
Flipping the elements seems to have flipped the sides. We still get the bottom/left sides right, what is different now is the existing 1D sides that were turned into 2D sides.

Figures show the case. The extrusion direction is in the depth of the figure. The left and right (brown, green) are the sidesets that got broken. left is reference
Screen Shot 2022-11-09 at 8 03 15 AM
Screen Shot 2022-11-09 at 8 03 32 AM

I think it s just a convention in the nodes selected for the swaps at this point?
The swaps selected in AdvancedExtruder 'preserved' these sidesets on a Z-axis extrusion.

**Comment by roystgnr** [comment]
>Left and right exist as 1D sidesets, and becomes 2D sidesets in the extruded mesh, probably because the nodes are in the relevant nodeset and the sideset is created off of this nodeset

Ah!  I swapped the `BoundaryInfo` side entries but not edge or node entries!  Thanks for catching that.

**Comment by GiudGiud** [comment]
the nodeset seems wrong, and the sideset for that nodeset is wrong as well
I m plotting both left/right sidesets but only one (right) nodeset. 

Screen Shot 2022-11-09 at 9 57 50 AM

**Comment by roystgnr** [comment]
Can you point me to the branch + input file to reproduce this?

**Comment by GiudGiud** [comment]
https://github.com/RocksSavage/moose/tree/create-SymmetryTransformerGenerator
the input is
test/tests/meshgenerators/advanced_extruder_generator/elem_flip.i

I complexified it locally but it s not needed to show the problem

**Comment by GiudGiud** [comment]
is this still correct with an element flip?
It s copying the old boundary id to the new mesh
```
            // For 2D->3D extrusion, we give the boundary IDs
            // for side s on the old element to side s+1 on the
            // new element.  This is just a happy coincidence as
            // far as I can tell...
            if (_boundary_swap_pairs.empty())
              boundary_info.add_side(new_elem, cast_int<unsigned short>(s + 1), ids_to_copy);
```

**Comment by roystgnr** [comment]
Okay, that is the problem.  That `add_side()` would be correct if it were done *before* the flip, but in the branch it's being done *after* the flip.  So we flip the element, the sides identified by each side number change accordingly, but then we add boundary info to the side number that would have been appropriate for the original inverted element, not the side number that's appropriate for the flipped element.

Why do `volume()` tests and an individual `flip()` call on each element?  We ought to be able to just `orient_elements()` once at the end before returning.  Shorter code, wouldn't have this bug, the test for inversion in `orient()` (and thereby in `orient_elements()` is cheaper than `volume()` ... and possibly more future-proofed, if we ever make `volume()` return an absolute value on cells the way it currently does on faces.

**Comment by GiudGiud** [comment]
doing it that other way works. 

Thanks for looking into it.

**Pull request merged**